Mathlib/RingTheory/Subring/Basic.lean

```diff
@@ -0,0 +1,1485 @@
+
+* `map f A : subring B` : the image of a subring `A` along the ring homomorphism `f`.
+
+* `prod A B : subring (R × S)` : the product of subrings
+
+* `f.range : subring B` : the range of the ring homomorphism `f`.
+
+* `eq_locus f g : subring R` : given ring homomorphisms `f g : R →+* S`,
+     the subring of `R` where `f x = g x`
+
+## Implementation notes
+
+A subring is implemented as a subsemiring which is also an additive subgroup.
+The initial PR was as a submonoid which is also an additive subgroup.
+
+Lattice inclusion (e.g. `≤` and `⊓`) is used rather than set notation (`⊆` and `∩`), although
+`∈` is defined as membership of a subring's underlying set.
+
+## Tags
+subring, subrings
+-/
+
+
+-- Porting note: currently global - open BigOperators
+
+universe u v w
+
+variable {R : Type u} {S : Type v} {T : Type w} [Ring R]
+
+section SubringClass
+
+/-- `subring_class S R` states that `S` is a type of subsets `s ⊆ R` that
+are both a multiplicative submonoid and an additive subgroup. -/
+class SubringClass (S : Type _) (R : outParam <| Type u) [Ring R] [SetLike S R] extends
+  SubsemiringClass S R, NegMemClass S R : Prop
+#align subring_class SubringClass
+
+-- See note [lower instance priority]
+instance (priority := 100) SubringClass.addSubgroupClass (S : Type _) (R : outParam <| Type u)
+    [SetLike S R] [Ring R] [h : SubringClass S R] : AddSubgroupClass S R :=
+  { h with }
+#align subring_class.add_subgroup_class SubringClass.addSubgroupClass
+
+variable [SetLike S R] [hSR : SubringClass S R] (s : S)
+
+theorem coe_int_mem (n : ℤ) : (n : R) ∈ s := by simp only [← zsmul_one, zsmul_mem, one_mem]
+#align coe_int_mem coe_int_mem
+
+namespace SubringClass
+
+instance (priority := 75) toHasIntCast : IntCast s :=
+  ⟨fun n => ⟨n, coe_int_mem s n⟩⟩
+#align subring_class.to_has_int_cast SubringClass.toHasIntCast
+
+-- Prefer subclasses of `ring` over subclasses of `subring_class`.
+/-- A subring of a ring inherits a ring structure -/
+instance (priority := 75) toRing : Ring s :=
+  Subtype.coe_injective.ring (↑) rfl rfl (fun _ _ => rfl) (fun _ _ => rfl) (fun _ => rfl)
+    (fun _ _ => rfl) (fun _ _ => rfl) (fun _ _ => rfl) (fun _ _ => rfl) (fun _ => rfl) fun _ => rfl
+#align subring_class.to_ring SubringClass.toRing
+
+-- Prefer subclasses of `ring` over subclasses of `subring_class`.
+/-- A subring of a `comm_ring` is a `comm_ring`. -/
+instance (priority := 75) toCommRing {R} [CommRing R] [SetLike S R] [SubringClass S R] :
+    CommRing s :=
+  Subtype.coe_injective.commRing (↑) rfl rfl (fun _ _ => rfl) (fun _ _ => rfl) (fun _ => rfl)
+    (fun _ _ => rfl) (fun _ _ => rfl) (fun _ _ => rfl) (fun _ _ => rfl) (fun _ => rfl) fun _ => rfl
+#align subring_class.to_comm_ring SubringClass.toCommRing
+
+-- Prefer subclasses of `ring` over subclasses of `subring_class`.
+/-- A subring of a domain is a domain. -/
+instance (priority := 75) {R} [Ring R] [IsDomain R] [SetLike S R] [SubringClass S R] : IsDomain s :=
+  NoZeroDivisors.to_isDomain _
```

Comment:
This seems to time out over the unification problem `Semiring.toNonUnitalSemiring.1 =?= (SubsemiringClass.toNonAssocSemiring s).1`. I blame leanprover/lean4#2003 :(

Reply:
2003 is merged :octopus: